### PR TITLE
Extract cli package

### PR DESCRIPTION
**Pull request by izumin5210**

### CODE REVIEW

File: pkg/cli/ui.go [hunk 1]
```diff
@@ -1,13 +1,105 @@
...
+
+type uiImpl struct {
+	out       io.Writer
+	inSection bool
+	inputUI   *input.UI
+}
+
+func (u *uiImpl) Section(msg string) {
+	if u.inSection {
+		fmt.Fprintln(u.out)
```
Comment (watchbunny) [inline pkg/cli/ui.go]: **[errcheck]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	fmt.Fprintln(u.out)

File: pkg/cli/ui.go [hunk 2]
```diff
@@ -1,13 +1,105 @@
...
+		u.inSection = false
+	}
+	printTypeSection.Fprintln(u.out, msg)
+}
+
+func (u *uiImpl) Subsection(msg string) {
+	if u.inSection {
+		fmt.Fprintln(u.out)
```
Comment (watchbunny) [inline pkg/cli/ui.go]: **[errcheck]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	fmt.Fprintln(u.out)

File: pkg/excmd/executor.go [hunk 3]
```diff
@@ -93,16 +85,16 @@ func (e *executor) exec(c *Command, cmd *exec.Cmd) (out []byte, err error) {
 		wg.Add(2)
 		go func() {
 			defer wg.Done()
-			io.Copy(e.outWriter, io.TeeReader(outReader, &buf))
+			io.Copy(e.io.Out, io.TeeReader(outReader, &buf))
```
Comment (watchbunny) [inline pkg/excmd/executor.go]: **[errcheck]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	io.Copy(e.io.Out, io.TeeReader(outReader, &buf))

File: pkg/excmd/executor.go [hunk 4]
```diff
@@ -93,16 +85,16 @@ func (e *executor) exec(c *Command, cmd *exec.Cmd) (out []byte, err error) {
...
 		}()
 		closers = append(closers, outReader.Close)
 		go func() {
 			defer wg.Done()
-			io.Copy(e.errWriter, io.TeeReader(errReader, &buf))
+			io.Copy(e.io.Err, io.TeeReader(errReader, &buf))
```
Comment (watchbunny) [inline pkg/excmd/executor.go]: **[errcheck]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	io.Copy(e.io.Err, io.TeeReader(errReader, &buf))

File: pkg/grapicmd/cmd/version.go [hunk 5]
```diff
@@ -17,12 +17,13 @@ func newVersionCommand(ctx *grapicmd.Ctx) *cobra.Command {
...
-			if ctx.Prebuilt {
-				buf.WriteString(" (" + ctx.BuildDate + " " + ctx.Revision + ")")
+			b := ctx.Build
+			buf := bytes.NewBufferString(b.AppName + " " + b.Version)
+			if b.Prebuilt {
+				buf.WriteString(" (" + b.BuildDate + " " + b.Revision + ")")
 			}
 			buf.WriteString("\n")
-			fmt.Fprintf(ctx.OutWriter, buf.String())
+			fmt.Fprintf(ctx.IO.Out, buf.String())
```
Comment (watchbunny) [inline pkg/grapicmd/cmd/version.go]: **[errcheck]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	fmt.Fprintf(ctx.IO.Out, buf.String())

File: pkg/protoc/wrapper.go [hunk 6]
```diff
@@ -63,15 +61,15 @@ func (e *wrapperImpl) installPlugins(ctx context.Context) error {
...
+	protoFiles, err := e.cfg.ProtoFiles(e.fs, e.rootDir.String())
 	if err != nil {
 		return errors.WithStack(err)
 	}
 
 	var errs []error
 	for _, path := range protoFiles {
 		err = e.execProtoc(ctx, path)
-		relPath, _ := filepath.Rel(e.rootDir, path)
+		relPath, _ := filepath.Rel(e.rootDir.String(), path)
```
Comment (watchbunny) [inline pkg/protoc/wrapper.go]: **[errcheck]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	relPath, _ := filepath.Rel(e.rootDir.String(), path)

File: pkg/cli/nop_ui.go [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package cli
+
+var NopUI = &nopUIImpl{}
```
Comment (watchbunny) [inline pkg/cli/nop_ui.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported var NopUI should have comment or be unexported

File: pkg/cli/ui.go [hunk 8]
```diff
@@ -1,13 +1,105 @@
...
+		ValidateFunc: func(ans string) error {
+			clog.Debug("receive user input", "query", msg, "input", ans)
+			if ans != "Y" && ans != "n" {
+				return fmt.Errorf("input must be Y or n")
+			}
+			return nil
+		},
+	})
+	if err != nil {
+		return false, err
```
Comment (watchbunny) [inline pkg/cli/ui.go]: **[wraperr]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
	github.com/izumin5210/grapi/pkg/cli.Confirm

File: pkg/cli/types.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	colorable "github.com/mattn/go-colorable"
+)
+
+type IO struct {
```
Comment (watchbunny) [inline pkg/cli/types.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type IO should have comment or be unexported

File: pkg/cli/types.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	In  io.Reader
+	Out io.Writer
+	Err io.Writer
+}
+
+func DefaultIO() *IO {
```
Comment (watchbunny) [inline pkg/cli/types.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function DefaultIO should have comment or be unexported

File: pkg/cli/types.go [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+		Err: os.Stderr,
+	}
+	if runtime.GOOS == "windows" {
+		io.Out = colorable.NewColorableStdout()
+		io.Err = colorable.NewColorableStderr()
+	}
+	return io
+}
+
+type RootDir string
```
Comment (watchbunny) [inline pkg/cli/types.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type RootDir should have comment or be unexported

File: pkg/cli/types.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+func (d RootDir) String() string { return string(d) }
+
+func (d RootDir) Join(elem ...string) string {
```
Comment (watchbunny) [inline pkg/cli/types.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method RootDir.Join should have comment or be unexported

File: pkg/cli/types.go [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	return filepath.Join(append([]string{d.String()}, elem...)...)
+}
+
+func (d RootDir) BinDir() string {
```
Comment (watchbunny) [inline pkg/cli/types.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method RootDir.BinDir should have comment or be unexported

File: pkg/cli/ui.go [hunk 14]
```diff
@@ -1,13 +1,105 @@
...
 	"io"
 	"strconv"
+	"sync"
 
 	"github.com/fatih/color"
+	"github.com/izumin5210/clicontrib/pkg/clog"
+	"github.com/tcnksm/go-input"
 )
 
+type UI interface {
```
Comment (watchbunny) [inline pkg/cli/ui.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type UI should have comment or be unexported

File: pkg/cli/ui.go [hunk 15]
```diff
@@ -1,13 +1,105 @@
...
+	ItemFailure(msg string)
+	Confirm(msg string) (bool, error)
+}
+
+var (
+	ui   UI
+	uiMu sync.Mutex
+)
+
+func UIInstance(io *IO) UI {
```
Comment (watchbunny) [inline pkg/cli/ui.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function UIInstance should have comment or be unexported

File: pkg/grapicmd/context.go [hunk 16]
```diff
@@ -45,51 +36,67 @@ type Config struct {
 	}
 }
 
+type BuildConfig struct {
```
Comment (watchbunny) [inline pkg/grapicmd/context.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type BuildConfig should have comment or be unexported

File: pkg/grapicmd/context.go [hunk 17]
```diff
@@ -45,51 +36,67 @@ type Config struct {
...
 	err = c.Viper.UnmarshalKey("protoc", &c.ProtocConfig)
 	if err != nil {
+		clog.Warn("failed to parse protoc config", "error", err)
 		return errors.WithStack(err)
 	}
 
 	return nil
 }
+
+func (c *Ctx) IsInsideApp() bool {
```
Comment (watchbunny) [inline pkg/grapicmd/context.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method Ctx.IsInsideApp should have comment or be unexported

File: pkg/grapicmd/di/providers.go [hunk 18]
```diff
@@ -19,57 +19,45 @@ import (
...
 )
 
-func ProvideUI(ctx *grapicmd.Ctx) clui.UI {
-	uiMu.Lock()
-	defer uiMu.Unlock()
-	if ui == nil {
-		ui = clui.New(ctx.OutWriter, ctx.InReader)
-	}
-	return ui
+func ProvideIO(ctx *grapicmd.Ctx) *cli.IO {
```
Comment (watchbunny) [inline pkg/grapicmd/di/providers.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function ProvideIO should have comment or be unexported

File: pkg/grapicmd/di/providers.go [hunk 19]
```diff
@@ -19,57 +19,45 @@ import (
...
+	return ctx.IO
 }
 
-func ProvideCommandExecutor(ctx *grapicmd.Ctx, ui clui.UI) excmd.Executor {
-	return excmd.NewExecutor(ctx.OutWriter, ctx.ErrWriter, ctx.InReader)
-}
-
-func ProvideGenerator(ctx *grapicmd.Ctx, ui clui.UI) module.Generator {
+func ProvideGenerator(ctx *grapicmd.Ctx, ui cli.UI) module.Generator {
```
Comment (watchbunny) [inline pkg/grapicmd/di/providers.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function ProvideGenerator should have comment or be unexported

File: pkg/grapicmd/di/providers.go [hunk 20]
```diff
@@ -90,30 +78,30 @@ func ProvideToolRepository(ctx *grapicmd.Ctx, gexCfg *gex.Config) (tool.Reposito
 	return toolRepo, nil
 }
 
-func ProvideProtocWrapper(ctx *grapicmd.Ctx, ui clui.UI, toolRepo tool.Repository) protoc.Wrapper {
+func ProvideProtocWrapper(ctx *grapicmd.Ctx, ui cli.UI, toolRepo tool.Repository) protoc.Wrapper {
```
Comment (watchbunny) [inline pkg/grapicmd/di/providers.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function ProvideProtocWrapper should have comment or be unexported

File: pkg/grapicmd/di/providers.go [hunk 21]
```diff
@@ -90,30 +78,30 @@ func ProvideToolRepository(ctx *grapicmd.Ctx, gexCfg *gex.Config) (tool.Reposito
...
 		ctx.Execer,
 		ui,
 		toolRepo,
 		ctx.RootDir,
-		ctx.BinDir,
 	)
 }
 
-func ProvideInitializeProjectUsecase(ctx *grapicmd.Ctx, gexCfg *gex.Config, ui clui.UI, generator module.Generator) usecase.InitializeProjectUsecase {
+func ProvideInitializeProjectUsecase(ctx *grapicmd.Ctx, gexCfg *gex.Config, ui cli.UI, generator module.Generator) usecase.InitializeProjectUsecase {
```
Comment (watchbunny) [inline pkg/grapicmd/di/providers.go]: **[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported function ProvideInitializeProjectUsecase should have comment or be unexported

File: pkg/cli/types.go (hunk 9 above)
Comment (izumin5210) [reply pkg/cli/types.go]: fixed in 3041c36

File: pkg/cli/types.go (hunk 10 above)
Comment (izumin5210) [reply pkg/cli/types.go]: fixed in 3041c36

File: pkg/cli/types.go (hunk 11 above)
Comment (izumin5210) [reply pkg/cli/types.go]: fixed in 3041c36

File: pkg/cli/types.go (hunk 12 above)
Comment (izumin5210) [reply pkg/cli/types.go]: fixed in 3041c36

File: pkg/cli/types.go (hunk 13 above)
Comment (izumin5210) [reply pkg/cli/types.go]: fixed in 3041c36

File: pkg/cli/ui.go (hunk 14 above)
Comment (izumin5210) [reply pkg/cli/ui.go]: fixed in 3041c36

File: pkg/cli/ui.go (hunk 15 above)
Comment (izumin5210) [reply pkg/cli/ui.go]: fixed in 3041c36

File: pkg/grapicmd/context.go (hunk 16 above)
Comment (izumin5210) [reply pkg/grapicmd/context.go]: fixed in 3041c36

File: pkg/grapicmd/context.go (hunk 17 above)
Comment (izumin5210) [reply pkg/grapicmd/context.go]: fixed in 3041c36

**Pull request merged**